### PR TITLE
router: add new path-splitting prefix matcher

### LINKED ISSUE #18148: Add efficient path-based prefix matching
*Title*: Add efficient path-based prefix matching

*Description*:
The Kubernetes Ingress spec, and soon the gateway-api spec (https://github.com/kubernetes-sigs/gateway-api/issues/866) implement a different form of prefix path matching than Envoy. From the spec:

> Prefix: Matches based on a URL path prefix split by /. Matching is case sensitive and done on a path element by element basis. A path element refers to the list of labels in the path split by the / separator. A request is a match for path p if every p is an element-wise prefix of p of the request path.

Essentially, if we want to match `prefix=/foo` in this manner we have two options:
* Convert to a regex match (like https://github.com/projectcontour/contour/blob/2b3376449bedfea7b8cea5fbade99fb64009c0f6/internal/envoy/v3/route.go#L78)
* Create two routes, one with `prefix=/foo/` and one with `exact=/foo`.

The first option has a runtime cost in route matching. The second has a cost in Envoy+Control plane during XDS processing.

Given how common this is/will be in Envoy users, it would be great if there was a more efficient way to control this. For example, setting a bool var like `case_sensitive` to control the behavior or defining some new type of `path_specifier`.

cc @jpeach @youngnick @lambdai

### PR DESCRIPTION
**Pull request by tpetkov-VMW**
Commit Message: Extend RouteMatch with a new field for subpath matching
Additional Description: The new field would allow more efficient generation of routes, replacing pairs of path+prefix routes into one path_separated_prefix route
Risk Level: Low
Testing: Unit test
Docs Changes: inline
Release Notes: Added
Fixes #18148

### CODE REVIEW

File: source/common/router/config_impl.cc
Comment (lambdai) [reply source/common/router/config_impl.cc:1432]: reuse that function to work with "/abc#anchor". Envoy allow anchor in some settings

**Review by lambdai — CHANGES_REQUESTED** [review]
Thank you! Almost there

**Comment by tpetkov-VMW** [comment]
Not sure about the naming, `path_separated_prefix` seems long.
Thinking about `subpath` as an alternative

**Review by lambdai — APPROVED** [review]
Thank you! The test cases are awesome!

**Comment by htuch** [comment]
I'm on the fence here on whether to do this vs the other alternatives in https://github.com/envoyproxy/envoy/issues/18148. @kyessenov  @snowp

**Comment by kyessenov** [comment]
Until we can use the matcher API to select routes, I don't see another way to support this without adding API fields. So I think the stumbling block is moving towards matchers in RDS, and I'm not up to date on that area.

**Comment by htuch** [comment]
@snowp can you help out here with matcher + RouteConfiguration?

**Comment by snowp** [comment]
It is possible today to use the new matching structure via RDS (https://github.com/envoyproxy/envoy/pull/17633), so if that was the main perceived blocker perhaps this should go into the new matcher framework? Or are there other reasons to want to continue to add to the legacy route matcher?

**Comment by tpetkov-VMW** [comment]
> It is possible today to use the new matching structure via RDS (#17633), so if that was the main perceived blocker perhaps this should go into the new matcher framework? Or are there other reasons to want to continue to add to the legacy route matcher?

Adding it to the legacy route matcher would give users a more easily consumable feature.
Frankly, the new matching structure is still confusing to me, and that's the main reason I chose the legacy routes.
I could try to open another PR for the RDS version and close this one, or perhaps they could coexist?

**Comment by phlax** [comment]
@tpetkov-VMW this is failing on version history formatting - but i guess you are waiting for feedback before proceeding/fixing

/wait-any

**Comment by tpetkov-VMW** [comment]
I'm not sure what to do, and whether this patch is desired at all.
I'll fix what's needed to pass the checks, so the PR is ready to go.

**Comment by htuch** [comment]
@tpetkov-VMW I've discussed this with @envoyproxy/api-shepherds and we're in-principle good with moving ahead with this PR. The consensus is that we will continue to support and evolve the existing matchers (and ideally integrate them as leaf functions into the generalized match trees).

File: api/envoy/config/route/v3/route_components.proto
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:519]:
```suggestion
    // *path_separated_prefix* or have it as a prefix, followed by '/'
```

File: api/envoy/config/route/v3/route_components.proto
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:520]: Maybe give some illustrative match / no match examples.

**Review by htuch — CHANGES_REQUESTED** [review]

File: api/envoy/config/route/v3/route_components.proto
Comment (phlax) [reply api/envoy/config/route/v3/route_components.proto:519]: please dont use `*` for inline literals - use double backticks ` `` `

same elsewhere in this PR where you have used `*` or `'`

Comment (tpetkov-VMW) [reply api/envoy/config/route/v3/route_components.proto:519]: Will do, is there a guide for the formatting of comments?
I might look things over, to save reviewers some work.

Comment (phlax) [reply api/envoy/config/route/v3/route_components.proto:519]: sort of - here https://github.com/envoyproxy/envoy/blob/main/api/CONTRIBUTING.md#documentation-guidelines

originally the guidance suggested using single backticks - and the codebase still has a lot of literals with either single backticks or asterisks

for reference here is the commit where i flipped the advice to use (the more semantically correct) backticks for literals

here https://github.com/envoyproxy/envoy/commit/eb18d71972e1f8af714d7843eb206108aad153c7

File: api/envoy/config/route/v3/route_components.proto
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:517]:
```suggestion
    // If specified, the route is a path-separated prefix rule meaning that the
```

**Review by htuch — COMMENTED** [review]
/lgtm api

File: source/common/router/config_impl.h
Comment (adisuissa) [inline source/common/router/config_impl.h:1117]: Looking at coverage: https://storage.googleapis.com/envoy-pr/2b4b01e/coverage/source/common/router/config_impl.h.gcov.html it seems that these lines are not covered.
Can you add the similar tests as for `PrefixRouteEntryImpl`.

File: source/common/router/config_impl.cc
Comment (adisuissa) [inline source/common/router/config_impl.cc:1426]: Looking at coverage: https://storage.googleapis.com/envoy-pr/2b4b01e/coverage/source/common/router/config_impl.cc.gcov.html this line isn't covered. Can you add similar coverage as in `PrefixRouteEntryImpl`?

**Review by adisuissa — COMMENTED** [review]
Thanks, looks good!
Left a few nits.
/wait

**Review by adisuissa — COMMENTED** [review]
Thanks, LGTM!
/lgtm api
LGTM code changes.
Can you please add release-notes?

**Comment by tpetkov-VMW** [comment]
> Can you please add release-notes?
I've added them in /docs/root/version_history/current.rst but not sure what to write in the comment at the top
Is just "inline" okay?

**Comment by adisuissa** [comment]
> > Can you please add release-notes?
> 
> I've added them in /docs/root/version_history/current.rst but not sure what to write in the comment at the top Is just "inline" okay?

Yep. I typically write "Added", unless there's some special doc that was added.

BTW: please merge main.

File: source/common/router/config_impl.cc
Comment (soulxu) [inline source/common/router/config_impl.cc:1437]: The `path_patcher_` will did the invoke `Http::PathUtil::removeQueryAndFragment` on the path again.
https://github.com/envoyproxy/envoy/blob/b77651d0ded885f085f0ac804d3a5f339a9b1925/source/common/common/matchers.cc#L126

Not sure whether worth to use `envoy::type::matcher::v3::StringMatcher` here directly. Leave this to other reviewers.

**Comment by soulxu** [comment]
LGTM

File: source/common/router/config_impl.cc
Comment (adisuissa) [reply source/common/router/config_impl.cc:1437]: Yes, you are correct.
I think the alternative is to not invoke `path_matcher_->match` method, which could result in inconsistencies in the future if this method is updated.

**Review by adisuissa — APPROVED** [review]
LGTM thanks!
Needs approval from senior-maintainer.

File: source/common/router/config_impl.cc
Comment (tpetkov-VMW) [reply source/common/router/config_impl.cc:1437]: It seemed like the most concise way to do it, without expanding the StringMatcher definition.

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (htuch) [inline source/extensions/filters/http/jwt_authn/matcher.cc:170]: I'm a bit unhappy we have duplicative logic here @qiwzhang. I've probably complained before and this is unrelated to this PR, so just an observation, but by having to copy+paste all our matching logic in two places, we are leaving open a significant opportunity for drift and check vs. use semantic differences between this and routing.

File: test/common/router/config_impl_test.cc
Comment (htuch) [inline test/common/router/config_impl_test.cc:7830]: Nit: make all these const

File: test/extensions/filters/http/jwt_authn/matcher_test.cc
Comment (htuch) [inline test/extensions/filters/http/jwt_authn/matcher_test.cc:186]: It would be good to use consistent test cases that are easy to correlate across the JWT and core router.

**Review by htuch — APPROVED** [review]
Looks good, my main concern is not specific to this PR but that we are having to constantly copy+paste our match logic across this and the JWT filter. This is related to the generic matcher discussion @envoyproxy/api-shepherds @snowp, since if we were using that, we would be able to just implement this once.

File: test/extensions/filters/http/jwt_authn/matcher_test.cc
Comment (tpetkov-VMW) [reply test/extensions/filters/http/jwt_authn/matcher_test.cc:186]: Do you mean "equalize the test inputs per test case"?
I'll start with that, and if more work needs to be done - just elaborate on it.

Comment (htuch) [reply test/extensions/filters/http/jwt_authn/matcher_test.cc:186]: Not sure about the definition of equalize :) Anything you can do to make tests more consistent between JWT and core router. Since you're taking on #20578 I'm fine with punting to that if you prefer to do it there.

Comment (tpetkov-VMW) [reply test/extensions/filters/http/jwt_authn/matcher_test.cc:175]: I made it so the jwt test cases are a subset of the router ones, and the test strings are pretty much the same.
They don't look exactly the same, but the correlation should be obvious.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
